### PR TITLE
Graph RNNT: Grid- and Compose-Transducer. W-Transducer loss

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do?

Implement Grid- and Compose-Transducer along with W-Transducer loss according to the paper "Powerful and Extensible WFST Framework for Rnn-Transducer Losses" https://ieeexplore.ieee.org/document/10096679 (https://arxiv.org/abs/2303.10384)

**Collection**: [ASR]

# Changelog 
- `GraphTransducerLossBase` abstract class  with the interface for Graph-based loses
- RNN-T implementation in `GraphRnntLoss` with tests
- W-Transducer implementation in `GraphWTransducerLoss` with tests
- add `GraphRnntLoss` + `GraphWTransducerLoss` to RNN-T loss resolver

# Usage

```shell
# training some model with Graph-RNNT loss
python examples/asr/asr_transducer/speech_to_text_rnnt_bpe.py \
    <other params> \
    model.loss.loss_name=graph_rnnt \
    ++model.loss.graph_rnnt_kwargs.double_scores=true

# training some model with W-Transducer loss
python examples/asr/asr_transducer/speech_to_text_rnnt_bpe.py \
    <other params> \
    model.loss.loss_name=graph_w_transducer \
    ++model.loss.graph_w_transducer_kwargs.eps_weight=0

```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/k2/w_transducer.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        Get unit schema (target text) graph for W-Transducer loss (Compose-Transducer).
+        Forward arcs represent text labels.
+
+        Example graph: text [1, 2], blank=0. Eps ids: 3, 4.
+
+        graph::
+
+                3:3:0                  0:0:1                  0:0:2
+              +-------+              +-------+              +-------+
+              v       |              v       |              v       |
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/w_transducer.py:85]: This ark should be `3:0:0` or `0:3:0`, shouldn't it? It is better not to have auxiliary indices in the final lattice, and so all those ark must become epsilon (0) in the end.

File: nemo/collections/asr/parts/k2/w_transducer.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+            +-----------+  1:1:0   +-----------+  2:2:1   +-----------+  -1:-1:-1  #===#
+            |     0     | -------> |     1     | -------> |     2     | ---------> H 3 H
+            +-----------+          +-----------+          +-----------+            #===#
+              ^ 0:0:0 |                                     ^ 4:4:2 |
+              +-------+                                     +-------+
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/w_transducer.py:90]: Same for `4:4:2`.

File: nemo/collections/asr/parts/k2/w_transducer.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+              v     |  |            v     |  |            v     | v v
+            +--------------+  0:0  +------------+  0:1   +------------+  0:2   +---+  -1:-1   #===#
+            |    0         | ----> |    1       | -----> |    2       | -----> | 3 | -------> H 4 H
+            +--------------+       +------------+        +------------+        +---+          #===#
+              ^ 2:0 | | | |         ^ 2:1 |  ^            ^ 2:2 |  ^             ^
+              +-----+ | | |         +-----+  |            +-----+  |             |
+                      | | |     3:0          |                     |             |
+                      | | +------------------+     3:0             |             |
+                      | +------------------------------------------+    3:0      |
+                      +----------------------------------------------------------+
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/w_transducer.py:158]: The arc from states 0 to 3 is redundant. It will never be used as it will be eaten up during the composition.
It also makes it harder to understand the schema.

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        Args:
+            num_frames: length of the sequence (in frames)
+            vocab_size: number of labels (including blank)
+            device: device for tensor to construct
+
+        Returns:
+            temporal schema graph (k2.Fsa).
+            Labels: <unit>:<frame_index>. <unit> is a unit from vocab + special units (e.g., additional eps).
+        """
+        pass
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/graph_transducer.py:93]: Same

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+        Args:
+            units_tensor: tensor with target text
+            num_frames: length of the sequence (in frames)
+            vocab_size: number of labels (including blank)
+
+        Returns:
+            transducer lattice (k2.Fsa).
+            Labels: <unit>:<frame_index>:<unit_position> (k2.Fsa: labels, aux_labels, unit_positions)
+        """
+        pass
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/graph_transducer.py:109]: Same

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+
+        Args:
+            units_tensor: tensor with target text
+            vocab_size: number of labels (including blank). Needed to construct additional eps-arcs (in some cases).
+
+        Returns:
+            unit schema graph (k2.Fsa).
+            Labels: <unit>:<unit>:<unit_position> (k2.Fsa: labels, aux_labels, unit_positions)
+        """
+        pass
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/graph_transducer.py:77]: Maybe raise `NotImplementedError`?

File: nemo/collections/asr/parts/k2/w_transducer.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        vocab_size = logits.shape[-1]
+        target_fsas_vec = self.get_graphs_batched(logits_lengths, targets, target_lengths, vocab_size)
+
+        cast_context = force_float32_context() if self.cast_to_float32 else nullcontext()
+        with cast_context:
+            log_probs = F.log_softmax(logits, dim=-1)
+            with torch.no_grad():
+                indices = self.get_logits_indices(target_fsas_vec, logits.shape)
+                indices[target_fsas_vec.labels == -1] = 0
+                indices[target_fsas_vec.labels >= vocab_size] = 0  # eps
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/w_transducer.py:328]: Oh, I see. Well, anyway, it's better to address such things before `forward` and presume that the lattices are without redundancies at this point.

File: nemo/collections/asr/parts/k2/w_transducer.py (hunk 1 above)
Comment (GNroy) [reply nemo/collections/asr/parts/k2/w_transducer.py:85]: Upd: I found the place where you fix this. Please see that comment.

**Review by GNroy — DISMISSED** [review]
Overall LGTM, please see comments.

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+
+        Returns:
+            batched lattice - FsaVec (k2.Fsa)
+        """
+        batch_size = logits_lengths.shape[0]
+        with torch.no_grad():
+            if self.use_grid_implementation:
+                return k2.create_fsa_vec(
+                    [
+                        self.get_grid(
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:151]: It would be great, but it's beyond the current PR)

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+                        )
+                        for i in range(batch_size)
+                    ]
+                )
+
+            # composed version
+            text_fsas = [
+                self.get_unit_schema(units_tensor=targets[i, : target_lengths[i].item()], vocab_size=vocab_size,)
+                for i in range(batch_size)
+            ]
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:164]: Same)

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+            temporal_fsas = [
+                self.get_temporal_schema(
+                    num_frames=logits_lengths[i].item(), vocab_size=vocab_size, device=targets.device
+                )
+                for i in range(batch_size)
+            ]
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:170]: Same)

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+            target_fsas_vec = k2.compose(
+                k2.create_fsa_vec(text_fsas), k2.create_fsa_vec(temporal_fsas), treat_epsilons_specially=False
+            )
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:173]: For this PR let it be compose, since we are implementing Compose-Transducer)

File: nemo/collections/asr/parts/k2/w_transducer.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+    W-Transducer loss: RNN-T loss modification for training RNN-T model for the case
+    when some text at the beginning/end of the utterance is missing.
+    The resulting model behaves like the RNN-T model (no modification for decoding is required).
+    For details see "Powerful and Extensible WFST Framework for RNN-Transducer Losses" paper
+        https://ieeexplore.ieee.org/document/10096679
+    """
+
+    class LastBlankMode(PrettyStrEnum):
+        ALLOW_IGNORE = "allow_ignore"
+        FORCE_LAST = "force_last"
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/k2/w_transducer.py]: Done

File: nemo/collections/asr/parts/k2/w_transducer.py (hunk 1 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/w_transducer.py:85]: Won't fix for now

File: nemo/collections/asr/parts/k2/w_transducer.py (hunk 2 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/w_transducer.py:90]: Won't fix for now

File: nemo/collections/asr/parts/k2/graph_transducer.py (hunk 6 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:77]: `NotImplementedError` is redundant in abstract method

File: nemo/collections/asr/parts/k2/graph_transducer.py (hunk 5 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:109]: Same)

File: nemo/collections/asr/parts/k2/graph_transducer.py (hunk 4 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:93]: Same)

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+        cast_context = force_float32_context() if self.cast_to_float32 else nullcontext()
+        with cast_context:
+            log_probs = F.log_softmax(logits, dim=-1)
+            with torch.no_grad():
+                indices = self.get_logits_indices(target_fsas_vec, logits.shape)
+                indices[target_fsas_vec.labels == -1] = 0
+
+            # NB: do not assign scores -> modify, k2 will not update all scores correctly (modify -> assign)
+            scores = log_probs.flatten().index_select(-1, indices)
+            scores[target_fsas_vec.labels == -1] = 0
```
[the remark this answers is not in the document]
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py:479]: Added comment

File: nemo/collections/asr/parts/k2/w_transducer.py (hunk 7 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/w_transducer.py:330]: Added comments, as we discussed

File: nemo/collections/asr/parts/k2/w_transducer.py (hunk 3 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/w_transducer.py]: Removed this arc, also fixed construction code for temporal schema and tests

**Review by GNroy — APPROVED** [review]

File: nemo/collections/asr/losses/rnnt.py [hunk 14]
```diff
@@ -109,6 +119,20 @@ class RNNTLossConfig:
         is_available=True,
         installation_msg="Pure Pytorch implementation of Multiblank RNN-T loss. Slow and for debugging purposes only.",
     ),
+    "graph_w_transducer": RNNTLossConfig(
```
Comment (titu1994) [inline nemo/collections/asr/losses/rnnt.py:122]: Update the rest of the configs to explicitly show force fp32 is true

**Review by titu1994 — DISMISSED** [review]
Looks great. Minor comment about the new arg, make it explicit rather than a default. Easy to miss otherwise

File: nemo/collections/asr/losses/rnnt.py (hunk 14 above)
Comment (artbataev) [reply nemo/collections/asr/losses/rnnt.py:127]: Fixed (I updated all the configs)

**Review by GNroy — APPROVED** [review]
